File: bench/benches/tree.exs
Code.require_file("../lib/bench.exs", __DIR__)

# binary-trees allocation/GC stress (see bench/benches/tree.l). checksum = 2^D-1.
defmodule Tree do
  def mk(0), do: nil
  def mk(d), do: {mk(d - 1), mk(d - 1)}
  def ck(nil), do: 0
  def ck({l, r}), do: 1 + ck(l) + ck(r)
end

Bench.run("tree", fn -> Tree.ck(Tree.mk(16)) end)

File: bench/benches/tree.lua
-- binary-trees allocation/GC stress (see bench/benches/tree.l). checksum = 2^D-1.
package.path = (arg[0]:match("(.*/)") or "./") .. "../lib/?.lua;" .. package.path
local bench = require("bench")
local function mk(d) if d == 0 then return nil else return { mk(d - 1), mk(d - 1) } end end
local function ck(t) if t == nil then return 0 else return 1 + ck(t[1]) + ck(t[2]) end end
bench("tree", function() return ck(mk(16)) end)
